CI: Fix broken runners, update Fishtape & tests

SirWrexes opened the pull request:
The url was writing "404: Not found" in a file instead of Fishtape, and
the test runner was trying to interpret that.
The URL should now be poiting at the right file.

SirWrexes committed:
test: Update Fishtape and tests

all:
    Do not rely on $current_dirname to source helper.fish.
    Since helper.fish is the file that sets `current_dirname`,
    using this variable to find helper.fish makes no sense.

helper:
    - Simplify fetching of `current_dirname`
    - Set TESTNAME when sourced by parsing stack trace

with_repo:
    Enable setup and teardown

utils_test:
    Enable setup and teardown

__funcle_clean_test:
    Enable setup and teardown

__fundle_install_test:
    Enable setup and teardown

__fundle_plugin_test:
    Enable setup

__fundle_init_test:
    Collect __fundle_init output into a single string

__fundle_list_test:
    Enable setup
    Collect tokens into a single string in second test

__fundle_plugin_test:
    Collect tokens into a single string

__fundle_update_plugin:
    Use TESTNAME to show what file is being run

SirWrexes commented:
Looks like it brought up a lot more other issues. 😨

SirWrexes commented:
I see at commit #52f208c you changed the provider for Fishtape. Likely some compatibility issues came up. I'll check the log and see if I can help.

SirWrexes commented:
Tests should be fixed.
If they are and the Travis CI gives a positive result, I will squash the commits into one.
I will keep the list of changes in the commit description.

SirWrexes commented:
Okay so the Xenial runner with `FISH_PPA=release-2` fails because its version of Fish is too old and doesn't support some syntax that is used in Fishtape.
![image](https://user-images.githubusercontent.com/5567164/110251056-fa7c7780-7f7e-11eb-9fca-25f520a02629.png)
```
fish -c 'for file in test/*_test.fish; fishtape $file; end'
~/.config/fish/functions/fishtape.fish (line 53): Unsupported use of '||'. In fish, please use 'COMMAND; or COMMAND'.
                set --query argv[2] || set --append argv ""
                                     ^
from sourcing file ~/.config/fish/functions/fishtape.fish
	called on standard input
in command substitution
	called on standard input
source: Error while reading file “/home/travis/.config/fish/functions/fishtape.fish”
```

The MacOS one fails because it cannot find the Fish packet.
![image](https://user-images.githubusercontent.com/5567164/110251555-56480000-7f81-11eb-9cf3-2ec0b92880dc.png)
Also, in `prepare-ci.sh`, line 12, I see `brew instal` (missing one `l`).
I don't know if this is intentional, though it doesn't seem to cause any trouble.

And lastly, the runner using `FISH_PPA=nightly-master` surprisingly doesn't fail... But it still doesn't function properly. Its version of `git init` doesn't support the `-b` switch...
So a retro compatibility problem too.
![image](https://user-images.githubusercontent.com/5567164/110251589-737cce80-7f81-11eb-88cb-7174c999d9ea.png)

SirWrexes commented on Makefile:
```diff
@@ -1,6 +1,10 @@
 
-FISHTAPE = $(XDG_CONFIG_HOME)/fish/functions/fishtape.fish
+# In case someone wants to run the tests and already has Fishtape,
+# don't install it again, use the already installed one.
```
This is because I installed Fishtape via Fundle, so I kept having to remove `~/.config/fish/functions/fishtape.fish` ever time I ran `make test`.
I did try running tests without Fishtape installed via Fundle though, to make sure everything worked as intended.

jorgebucaran replied:
You need to Fish 3.0 to run Fishtape.

SirWrexes replied:
I suspected there was something of the sort.
But then doesn't that mean that we should also change the runner's settings ?
Also the comment box doesn't show this well but the line I changed here uses Fish `functions` to check the presence of Fishtape and find its path if one has it; or uses `$HOME/.config/fish/functions` as usual.

jorgebucaran replied:
There's no reason to not use at least Fish 3.0 IMO. [Here's how Fishtape tests itself with itself](https://github.com/jorgebucaran/fishtape/blob/e6e5fc23dd062ee5ed11828951bf2c85d6798db5/.github/workflows/ci.yml#L19-L21).

SirWrexes commented:
@jorgebucaran continung the discussion about runner settings and Fish version here...
I agree with you there is no reason not to use Fish >3.0.
So I suppose that would mean replacing `FISH_PPA=release-2` with `FISH_PPA=release-3` for the release test suite.
And I guess I should look into the docs of the Travis CI to see if it is possible to get some updated runners; with a working Fish package for the macOS one as well as a git version that supports the `-b` flag for the `init` subcommand in the Xenial one. :thinking: 
I suppose all of the Travis CI settings are located in `.travis.yml` right ?

SirWrexes committed:
ci: Update runner configuration

.travis.yml:
    Use "shell" as value for `language`
    Update/remove deprecated keys to suppress warnings/errors
    Enforce latest versions of the runners' OS
    Set git --depth to 1 to speed up jobs
    Use Fish >=3.0

run-tests.fish (NEW):
    Return non zero status when a test fails

Get rid of unnecessary prepare-ci.sh

SirWrexes commented:
Heck yeah !

Build jobs now work, and I've checked the logs, everything is in order !
The issue _was_ indeed a retro compatibility thing.

Travis CI uses Ubuntu Xenial 16.04 as well as macOS 13.0.1 by default, which are both quite old.
When testing apps with long term support, it makes sense, but since Fundle uses a [2020 git feature](https://github.com/git/git/commit/32ba12dab2acf1ad11836a627956d1473f6b851a) this is out of the question.
I enforced using the latest versions of the two and voilà.

Now I'll try to rewrite the Travis config a little more elegantly...
Mainly try to getting rid of  `make`, which doesn't fail a job even if some tests do fail.
Also maybe getting rid of `prepare-ci.sh` which doesn't really seem all that necessary.

After that, I'll cleanup the commit history and do a last push, and then this PR should be ready for merging !

SirWrexes commented:
Aaaaaand it's done ! Your tests are up to date and working perfectly.
Have a good day ! (ﾉ^ _^)ﾉ*:･ﾟ✧

danhper commented:
@SirWrexes I'm so sorry I never realized this PR was open, it must have slipped through my inbox. Thanks so much for fixing all this
